include/zephyr/toolchain/gcc.h

```diff
@@ -689,6 +689,13 @@ do {                                                                    \
 #define FUNC_NO_STACK_PROTECTOR
 #endif
 
+#ifndef __COUNTER__
+/* GCC versions before 4.3.0 do not support __COUNTER__
+ * but this should be good enough
+ */
+#define __COUNTER__ __LINE__
```

Comment:
`__LINE__` is not the same as `__COUNTER__`.

Especially not if `__COUNTER__` is used (directly or indirectly through other macros) inside a form of loop, cause in this case you can end up with duplicated names, like:
```
error: redeclaration of enumerator '__build_assert2643'
```
but counter would have given distinct values.

From docs:
> __COUNTER__
>
>    This macro expands to sequential integral values starting from 0. In conjunction with the ## operator, this provides a convenient means to generate unique identifiers.

`__LINE__` doesn't guarantee unique identifiers.